pkg/threadas/threadas_test.go

```diff
@@ -0,0 +1,152 @@
+//go:build darwin
+// +build darwin
+
+package threadas
+
+import (
+	"runtime"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const timeout = 100 * time.Millisecond
+
+func TestThreadAsNotRoot(t *testing.T) {
+	t.Parallel()
+
+	if syscall.Getuid() == 0 {
+		t.Skip("Skipping -- test requires not-root")
+	}
+
+	uids := getUidData()
+	fn := uids.GenerateTestFunc(t, "expected failure", assert.Equal)
+	require.Error(t, ThreadAs(fn, timeout, uint32(uids.Uid), uint32(uids.Gid)), "Fails when run as non-root")
+}
+
+func TestThreadAs(t *testing.T) {
+	t.Parallel()
+
+	if syscall.Getuid() != 0 {
+		t.Skip("Skipping -- test requires root")
+	}
+
+	targetUid := 501
+	targetGid := 20
+
+	targetUids := uidData{Uid: targetUid, Euid: targetUid, Gid: targetGid, Egid: targetGid}
+	fnTargetUidsEqual := targetUids.GenerateTestFunc(t, "matches target user", assert.Equal)
+	fnTargetUidsNotEqual := targetUids.GenerateTestFunc(t, "does not matches target user", assert.NotEqual)
+
+	myUids := getUidData()
+	fnMyUidsEqual := myUids.GenerateTestFunc(t, "matches my user", assert.Equal)
+	fnMyUidsNotEqual := myUids.GenerateTestFunc(t, "does not matches my user", assert.NotEqual)
+
+	// Be somewhat thoughtful in how parallel works here -- using
+	// t.Parallel will spawn a new thread, which potentially
+	// undermines some of what we're testing. But we do want the
+	// top level to be parallel, so that we have more thread
+	// churn.
+	for i := 1; i < 100; i++ {
```

Comment:
what is the purpose of running 100x? Just to make sure we don't get some kind of deadlocks?

Reply:
And no weird cross thread pollution. I updated the comment